### PR TITLE
[SPARK-31224][SQL] Add view support to SHOW CREATE TABLE

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

For now `SHOW CREATE TABLE` command doesn't support views, but `SHOW CREATE TABLE AS SERDE` supports it. Since the views syntax are the same between Hive DDL and Spark DDL, we should be able to support views in both two commands.

This is Hive syntax for creating views:

```
CREATE VIEW [IF NOT EXISTS] [db_name.]view_name [(column_name [COMMENT column_comment], ...) ]
  [COMMENT view_comment]
  [TBLPROPERTIES (property_name = property_value, ...)]
  AS SELECT ...;
```

https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-CreateView

This is Spark syntax for creating views:

```
CREATE [OR REPLACE] [[GLOBAL] TEMPORARY] VIEW [IF NOT EXISTS [db_name.]view_name
    create_view_clauses
    AS query;
```
https://spark.apache.org/docs/3.0.0-preview/sql-ref-syntax-ddl-create-view.html

Looks like it is the same. We could support views in both commands.

This patch proposes to add views support to `SHOW CREATE TABLE`.

### Why are the changes needed?

To extend the view support of `SHOW CREATE TABLE`, so users can use `SHOW CREATE TABLE` to show Spark DDL for views.

### Does this PR introduce any user-facing change?

Yes. `SHOW CREATE TABLE` can be used to show Spark DDL for views.

### How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
cc @cloud-fan @dongjoon-hyun @HyukjinKwon

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 1]
```diff
@@ -1109,21 +1138,37 @@ case class ShowCreateTableCommand(table: TableIdentifier)
...
+          tableMetadata
+        } else {
+          convertTableMetadata(tableMetadata)
+        }
       }
 
-      val stmt = showCreateDataSourceTable(metadata)
+      val builder = StringBuilder.newBuilder
+
+      val stmt = if (tableMetadata.tableType == VIEW) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: can we create a method to generate `CREATE VIEW`? So we can just call one method in the `if` branch.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 2]
```diff
@@ -1066,6 +1066,35 @@ trait ShowCreateTableCommandBase {
   protected def concatByMultiLines(iter: Iterable[String]): String = {
     iter.mkString("(\n  ", ",\n  ", ")\n")
   }
+
+  protected def showViewDataColumns(metadata: CatalogTable, builder: StringBuilder): Unit = {
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: Is this simple move?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala]: +1 for @cloud-fan 's advice. We had better have a function similar to `showCreateDataSourceTable`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:1070]: Yes.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:1164]: Created a method like suggestion.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala [hunk 3]
```diff
@@ -1109,21 +1145,33 @@ case class ShowCreateTableCommand(table: TableIdentifier)
...
-
         if ("true".equalsIgnoreCase(tableMetadata.properties.getOrElse("transactional", "false"))) {
           throw new AnalysisException(
             "SHOW CREATE TABLE doesn't support transactional Hive table. " +
               s"Please use `SHOW CREATE TABLE ${tableMetadata.identifier} AS SERDE` " +
               "to show Hive DDL instead.")
         }
 
-        convertTableMetadata(tableMetadata)
+        if (tableMetadata.tableType == VIEW) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:1155]: I think we should put it in a higher level
```
      val metadata = if (DDLUtils.isDatasourceTable(tableMetadata) || tableMetadata.tableType == VIEW) {
        tableMetadata
      } else {
        ...
```

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 3 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:1155]: We need to check for unsupported feature like `partitioned view` in this branch.

**Comment by viirya** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala:1155]: ah makes sense

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Pull request closed without merging**